californium-core/src/main/java/org/eclipse/californium/core/network/BaseMatcher.java

```diff
@@ -158,7 +163,8 @@ protected final Exchange matchNotifyResponse(final Response response, final Corr
 		final Exchange.KeyToken idByToken = Exchange.KeyToken.fromInboundMessage(response);
 		Exchange exchange = null;
 
-		final Observation obs = observationStore.get(response.getToken());
+		final EndpointIdentifier sourceEndpoint = response.getSourceEndpoint();
```

Comment:
OK.

So the main difference seems to be, that I would just pass the `CorrelationContext `into the `ObserveStore`, and you would prefer to have a "IdentifierFactory" to convert that `CorrelationContext` ahead into a `EndpointIdentifier`.